### PR TITLE
cilium: populate wildcard src->dst policy for transparent encryption

### PR DESCRIPTION
**Pull request by jrfastab**
Policy/State rules are populated when we receive kvstore or k8s
events. On egress we must match the (dst_addr, spi) tuple via the xfrm
policy and then link this to a state using the tmpl including the
same (dst_addr, spi). This results in a pair of policy/state rules
per node on the host. The above rules also specify a src address
so that we only match Cilium managed traffic. This includes the
IP address on cilium_host and the endpoint allocation range.

On the ingress side we have been doing the same but swapped. The
tuple (src, dst, spi) is used in the policy to match traffic and
then linked to the state via the tmpl. These are also added on
update events. However, this creates more rule than are actually
needed. By wildcarding the source on ingress traffic we can
match all traffic for the node with a single rule. Further the
existing scheme is buggy and dst-ip/spi pairs can collide.

The fix is to insert a wildcard decrypt rule when the local node
update event is received. Then on node update events only
add the egress rule.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7181)

### CODE REVIEW

**Comment by jrfastab** [comment]
Lets get CI test in then merge this so we get CI coverage.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 1]
```diff
@@ -33,6 +33,14 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type IPSecDir string
```
Comment (aanm) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:36]: For a moment I thought this was the "IPSecDirectory" and not "...Direction"

File: pkg/datapath/linux/node.go [hunk 2]
```diff
@@ -429,12 +429,18 @@ func (n *linuxNodeHandler) enableIPsec(newNode *node.Node) {
 		new4Net := &net.IPNet{IP: newNode.IPv4AllocCIDR.IP, Mask: newNode.IPv4AllocCIDR.Mask}
 		if newNode.IsLocal() {
 			n.replaceNodeIPSecInRoute(new4Net)
+			ciliumInternalIPv4 := newNode.GetCiliumInternalIP(false)
+			if ciliumInternalIPv4 != nil {
+				ipsecLocal := &net.IPNet{IP: n.nodeAddressing.IPv4().Router(), Mask: n.nodeAddressing.IPv4().AllocationCIDR().Mask}
+				ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP("0.0.0.0"), Mask: net.IPv4Mask(0, 0, 0, 0)}
```
Comment (aanm) [inline pkg/datapath/linux/node.go]: small nit, consider making this a "static" / "constant" variable in this file

File: pkg/datapath/linux/node.go [hunk 3]
```diff
@@ -449,15 +455,21 @@ func (n *linuxNodeHandler) enableIPsec(newNode *node.Node) {
 		if newNode.IsLocal() {
 			n.replaceHostRules()
 			n.replaceNodeIPSecInRoute(new6Net)
+			ciliumInternalIPv6 := newNode.GetCiliumInternalIP(true)
+			if ciliumInternalIPv6 != nil {
+				ipsecLocal := &net.IPNet{IP: n.nodeAddressing.IPv6().Router(), Mask: n.nodeAddressing.IPv6().AllocationCIDR().Mask}
+				ipsecIPv6Wildcard := &net.IPNet{IP: net.ParseIP("0::0"), Mask: net.CIDRMask(0, 0)}
```
Comment (aanm) [inline pkg/datapath/linux/node.go]: small nit, consider making this a "static" / "constant" variable in this file

**Review by aanm — COMMENTED** [review]
nothing blocking on my side

File: pkg/datapath/linux/node.go (hunk 2 above)
Comment (jrfastab) [reply pkg/datapath/linux/node.go]: will do.

File: pkg/datapath/linux/node.go (hunk 3 above)
Comment (jrfastab) [reply pkg/datapath/linux/node.go]: will do

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Added CI patch to this series on my test boxes with this patch CI tests pass reliably.

**Comment by jrfastab** [comment]
OK this is passing now. I'm going to run it a few times more to check for any flakes.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Lots of noise today but try again without net-next and see if we still hit the e1000e restart kernel bug.  2/3 passed on net-next with the failure appearing to be unrelated. Locally I ran 5/5 runs successfully.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Run time monitor test failed this seems unrelated will retry

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
OK removed failing patches until we get a resolution. For now pushing minimal fixes. And will create a new PR for the failing patch so we don't hold these up.

**Comment by jrfastab** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by jrfastab** [comment]
Tracking failure here https://github.com/cilium/cilium/issues/7282 this seems unrelated we don't even have encryption tests enabled here.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
FQDN flake also seems unrelated.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Passing now just to recount, these were seen and ignored assuming unrelated:
* https://github.com/cilium/cilium/issues/7282
* https://github.com/cilium/cilium/issues/7190
* ? https://github.com/cilium/cilium/issues/7128

**Review by tgraf — APPROVED** [review]

**Pull request merged**